Catalogue epic 1 3

Kolesnikov-Eugene opened the pull request:
Ссылка на командный таск-трекер:
https://github.com/users/Kolesnikov-Eugene/projects/2/views/1?layout=board

Архитектура проекта:  MVVM
Верстка UI: в коде
Декомпозиция проекта в CatalogDecomposition.md в корне проекта

Создан макет следующего экрана Collections c заглушкой для проверки функционала тапа по ячейке.
Реализация UI  и логики в следующей трети эпика.

Kolesnikov-Eugene committed:
CatalogViewController, CatalogView - UI ready.
CatalogTableViewCell - UI ready.
AlertPresenter and AlertPresenter model ready - to show alert when sort button in CatalogVC tapped.

Kolesnikov-Eugene committed:
Catalog service class created to interact with Network client and fetch data from server.
CatalogResult and Catalog structs to process data.
func configureCell created to display received data.
Using Combine to update state.
Filters are working properly.

Kolesnikov-Eugene committed:
Gradient animation added to collection cell while data is being loaded from server.
CatalogFilter storage - to keep last filter chosen by user.

Kolesnikov-Eugene committed:
Catalog screen functionality is ready.
Refactoring, code clean up.
CatalogAssembly class to build Catalog view.

Kolesnikov-Eugene committed:
UIBlockingProgressHUD class to show loading indicator while catalog data are being fetched from server.
Loading indicator stops animating when data is loaded and gradient animation of table view cell begins until photos for cells are loaded and presented.

Helvetii951 commented on FakeNFT/DesignSystem/Constants.swift:
```diff
@@ -33,4 +33,12 @@ struct Constants {
     static let addToBasketPicTitle = "basket_add"
     static let backwardPicTitle = "backward"
     static let addPlusButtonPicTitle = "add_plus_button"
+    
+    //MARK: - Catalog constants
+    static let catalogTableViewCellIdentifier = "catalogTableViewCell"
+    static let filterKey = "catalogFilter"
+    static let filterAlertTitle = "Сортировка"
+    static let filetNameButtonTitle = "По названию"
+    static let filterQuantityButtonTitle = "По количеству NFT"
+    static let closeAlertButtonTitle = "Отменить"
```
Такие свойства лучше разделять. Например, `catalogTableViewCellIdentifier` унести в ячейку. Выделить отдельную структуру для строк. 

А лучше всего для строк использовать библиотеки генерации ресурсов типа [SwiftGen](https://github.com/SwiftGen/SwiftGen).

Helvetii951 commented on FakeNFT/Scenes /Catalog/Cell/CatalogTableViewCell.swift:
```diff
@@ -0,0 +1,116 @@
+        startAnimation()
+        
+        descriptionLabel.text = "\(model.name) (\(model.nfts.count))"
+        
+        let processor = RoundCornerImageProcessor(cornerRadius: 12)
+        
```
`RoundCornerImageProcessor` из `Kingfisher` скругляет именно ту картинку, которую скачал. Т.е., если он скачал большую картинку(например, 2000х2000), то он скруглит у нее только 12 поинтов(такое число указано в коде проекта), что вряд ли будет заметно. `RoundCornerImageProcessor` никак не влияет на скругление UIImageView аватарки. Поэтому нужно скруглять саму `UIImageView`.

Helvetii951 commented on FakeNFT/Scenes /Catalog/Cell/CatalogTableViewCell.swift:
```diff
@@ -0,0 +1,116 @@
+    }
+    
+    //MARK: - Public methods
+    func configureCell(model: Catalog) {
+        UIBlockingProgressHUD.dismiss()
+        startAnimation()
+        
+        descriptionLabel.text = "\(model.name) (\(model.nfts.count))"
```
Скорее всего `UIBlockingProgressHUD.dismiss()` сюда попало по ошибке. Очень желательно убрать. Если не случайно, то такими делами нужно заниматься в контроллере, а не в ячейке.

Kolesnikov-Eugene replied:
Убрал пока что лоадер совсем на этапе загрузки картинок из сети. Оставил пока только анимацию ячеек при загрузке. Подумаю дальше над логикой показа лоадера, чтобы не дублировать его вместе с анимацией.

Helvetii951 commented on FakeNFT/Scenes /Catalog/Services/CatalogAssembly.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+final class CatalogAssembly {
+    static let shared = CatalogAssembly()
+    
+    func buildCatalogViewModel() -> CatalogViewModelProtocol {
+        let network = DefaultNetworkClient()
+        let service = CatalogService(networkClient: network)
+        let viewModel = CatalogViewModel(catalogService: service)
+        return viewModel
+    }
+}
```
Часто, при использовании экземпляра `static let shared` добавляют приватный инициализатор, реализуя тем самым паттерн [singleton](https://theswiftdev.com/swift-singleton-design-pattern/).

`buildCatalogViewModel` можно сделать статической. Тогда не придется создавать экземпляр `CatalogAssembly`.

Kolesnikov-Eugene replied:
Да, забыл про приватный инициализатор. Добавил везде, где требовалось

Helvetii951 commented on FakeNFT/Scenes /Catalog/Services/CatalogFilterStorage.swift:
```diff
@@ -0,0 +1,29 @@
+
+import Foundation
+
+final class CatalogFilterStorage {
+    static let shared = CatalogFilterStorage()
+    var filterDescriptor: String? {
+        get {
```
Тут тоже желательно добавить приватный инициализатор.

Helvetii951 commented on FakeNFT/Scenes /Catalog/View/CatalogView.swift:
```diff
@@ -0,0 +1,117 @@
+    }()
+    private var subscribes = [AnyCancellable]()
+    
+    init(frame: CGRect, viewModel: CatalogViewModelProtocol, delegate: CatalogViewDelegate) {
+        self.viewModel = viewModel
+        self.delegate = delegate
+        super.init(frame: frame)
+        setupUI()
+        bind()
+    }
```
👍 за инжект вью-модели по протоколу.

Helvetii951 commented on FakeNFT/Scenes /Catalog/View/CatalogViewController.swift:
```diff
@@ -0,0 +1,88 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        UIBlockingProgressHUD.show()
+    }
```
`UIBlockingProgressHUD.show()` лучше показывать/скрывать тогда, когда мы явно запускаем какую-то активность. Так лоадер может застрять на экране навсегда.

Helvetii951 commented on FakeNFT/Services/AlertPresenter.swift:
```diff
@@ -0,0 +1,35 @@
+
+import UIKit
+
+final class AlertPresenter {
+    func show(in vc: UIViewController, model: AlertModel) {
+        let alert = UIAlertController(
+            title: nil,
+            message: model.message,
+            preferredStyle: .actionSheet)
```
Функцию можно сделать статической.

Kolesnikov-Eugene committed:
Decomposition.md

Kolesnikov-Eugene committed:
Removed roundCornerImageProcessor (provided by kingfisher).
private init() in CatalogAssembly, AlertPresenter and CatalogFilterStorage.

Kolesnikov-Eugene commented:
Добавил к проекту файл с декомпозицией Decomposition.md, а также ссылку на таск-трекер (самый первый комментарий при создании пулл-реквеста).

Kolesnikov-Eugene committed:
coverUrl - optional in Catalog model

Kolesnikov-Eugene committed:
CatalogService func fetchCatalog() refactored

Kolesnikov-Eugene committed:
Network error alert in CatalogScreen (reload data on OK action).
Changed logic of updating state of CatalogViewModel.
Activity indicator while fetching data from server.

Kolesnikov-Eugene committed:
code cleaning

Kolesnikov-Eugene committed:
CatalogDecomposition

Kolesnikov-Eugene committed:
decomposition fix

Kolesnikov-Eugene committed:
Decomposition issues

Kolesnikov-Eugene commented:
@Helvetii951 
Ренат, привет! У нас тут какое-то недопонимание если честно по поводу работы моего проекта!
У меня все корректно работает, при любых настройках сети все загружает, при отсутствии вылетает Алерт для повторной загрузки! В чем проблема я не понимаю!
Ребята с моей команды и наставник проект себе скачивали, у них тоже все корректно работает!

Kolesnikov-Eugene committed:
SwiftLint disabled
Enable user sandboxing - NO

Kolesnikov-Eugene committed:
CatalogDecomomposition -> root
